### PR TITLE
Fail tests on `UserWarning: Detected call of lr_scheduler.step() before optimizer.step()`

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Preparation for a fix PR #11542 of an issue #5558.

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [n/a] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @akihironitta @rohitgr7

### CODE REVIEW

File: tests/core/test_lightning_optimizer.py [hunk 1]
```diff
@@ -195,6 +195,7 @@ def configure_optimizers(self):
     assert sgd_zero_grad.call_count == 10
 
 
+@pytest.mark.filterwarnings(r"ignore:Detected call of `lr_scheduler.step\(\)` before `optimizer.step\(\)`:UserWarning")
```
Comment (akihironitta) [inline tests/core/test_lightning_optimizer.py:198]: Since the optimizers are mocked in this test case, the actual `.step()` has never been called, and thus, it raises the warning. so, I'll just ignore the warning here.

File: tests/strategies/test_sharded_strategy.py [hunk 2]
```diff
@@ -16,6 +16,9 @@
     from fairscale.nn.data_parallel.sharded_ddp import ShardedDataParallel
 
 
+# TODO(akihironitta): Ignore the warning for now since lr_schedulers doesn't wrap new OSS optimizers that are
+# initialized by the strategy.
+@pytest.mark.filterwarnings(r"ignore:Detected call of `lr_scheduler.step\(\)` before `optimizer.step\(\)`:UserWarning")
```
Comment (akihironitta) [inline tests/strategies/test_sharded_strategy.py:21]: After `DDPShardedStrategy` creates new OSS optimizers at:
https://github.com/PyTorchLightning/pytorch-lightning/blob/5da065e287b44e2c1fe4f7951003813ed45365c9/pytorch_lightning/strategies/sharded.py#L71

lr_schedulers never get reinitialized , and thus, `lr_scheduler.optimizer._step_count` isn't incremented.

---

I think we should reinit lr_schedulers, too, after OSS optimizers are created in order to wrap the OSS optimizers' step method with a step counter at: https://github.com/pytorch/pytorch/blob/539acb29cd7a203b614034b5efd9f5144ff89ae9/torch/optim/lr_scheduler.py#L47-L72

File: tests/accelerators/test_ipu.py [hunk 3]
```diff
@@ -37,6 +37,11 @@
     import poptorch
 
 
+pytestmark = pytest.mark.filterwarnings(
+    r"ignore:Detected call of `lr_scheduler.step\(\)` before `optimizer.step\(\)`:UserWarning"
+)
+
+
```
Comment (akihironitta) [inline tests/accelerators/test_ipu.py:44]: Haven't looked into this further, but the reason for the warning could be a similar reason to the one in ddp_sharded as commented.

```python
FAILED tests/accelerators/test_ipu.py::test_all_stages[1] - UserWarning: Dete...
FAILED tests/accelerators/test_ipu.py::test_all_stages[4] - UserWarning: Dete...
FAILED tests/accelerators/test_ipu.py::test_stages_correct - UserWarning: Det...
FAILED tests/accelerators/test_ipu.py::test_autoreport - UserWarning: Detecte...
FAILED tests/accelerators/test_ipu.py::test_manual_poptorch_dataloader - User...
FAILED tests/accelerators/test_ipu.py::test_manual_poptorch_opts - UserWarnin...
FAILED tests/accelerators/test_ipu.py::test_manual_poptorch_opts_custom - Use...
FAILED tests/accelerators/test_ipu.py::test_default_opts - UserWarning: Detec...
```
https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=59459&view=logs&j=84dafd16-84ac-5ac7-3837-ae3525208541&t=dade7e55-b49c-56cd-3b84-da61e40b580f&l=12152

File: tests/core/test_lightning_optimizer.py (hunk 1 above)
Comment (awaelchli) [reply tests/core/test_lightning_optimizer.py:198]: can you add a brief comment there in the test about the reason for skipping? :)

**Review by awaelchli — APPROVED** [review]

File: tests/accelerators/test_ipu.py [hunk 4]
```diff
@@ -38,6 +38,13 @@
     import poptorch
 
 
+# TODO: Remove the warning filter. This is temporarily ignored because the warning seems always raised
```
[the remark this answers is not in the document]
Comment (akihironitta) [reply tests/accelerators/test_ipu.py:41]: We can address these TODOs before enabling the warning test (configured [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/a3853efba4e26ed527f2432fe246a1624754e76a/setup.cfg#L31)), but it'll need extra time and I think it's reasonable to enable the warning check to detect prospective failures in master even though some of the tests still fail.

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/accelerators/test_ipu.py:41]: I understand the sequencing you propose, but I am hesitant of adding these TODOs when they have very low priority and might just get forgotten.

You can always open a PR with the warning configuration to track which tests are currently affected without merging it to master.

[the remark this answers is not in the document]
Comment (akihironitta) [reply tests/accelerators/test_ipu.py:41]: > I am hesitant of adding these TODOs when they have very low priority and might just get forgotten.

Sure, I understand :]

> You can always open a PR with the warning configuration to track which tests are currently affected without merging it to master.

I might not get the workflow you have in mind. Do you mean I should add the warning configuration while working on a PR and remove it before merging it to master?

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/accelerators/test_ipu.py:41]: Yes, for example.

Or keep a separate PR (for example this one) just to check which are failing at the moment

[the remark this answers is not in the document]
Comment (akihironitta) [reply tests/accelerators/test_ipu.py:41]: Sure, sounds good :) I'll keep this PR open and work on the actual fix in separate PRs.
